fiftyone/core/odm/utils.py

```diff
@@ -492,62 +493,143 @@ def _parse_embedded_doc_fields(doc, dynamic):
-            if etype != ftype:
-                # @todo could provide an `add_mixed=True` option to declare
-                # mixed fields like this as a generic `Field`
-                fields_dict[name] = None
-            elif etype in (fof.ListField, fof.DictField):
-                if "subfield" in efield and efield["subfield"] != subfield:
-                    efield["subfield"] = None
-                elif subfield is not None:
-                    efield["subfield"] = subfield
-            elif ftype == fof.EmbeddedDocumentField:
-                _merge_embedded_doc_fields(efield["fields"], field["fields"])
-
-
-def _init_embedded_doc_fields(field):
-    fields_dict = {}
-    for _field in field["fields"]:
-        if _field["ftype"] == fof.EmbeddedDocumentField:
-            _init_embedded_doc_fields(_field)
+        if name not in merged or merged[name] is None:
+            merged[name] = field
+            if is_embedded_doc:
+                # Convert nested schemas to dicts once so recursive merges can
+                # be performed without repeatedly scanning lists.
+                field["fields"] = _to_field_mapping(field.get("fields", []))
+
+            continue
+
+        efield = merged[name]
+        etype = efield["ftype"]
+        etype_is_embedded = inspect.isclass(etype) and issubclass(
+            etype, fof.EmbeddedDocumentField
+        )
 
-        fields_dict[_field["name"]] = _field
+        if etype != ftype:
+            # @todo could provide an `add_mixed=True` option to declare mixed
+            # fields like this as a generic `Field`
+            merged[name] = None
+        elif etype in (fof.ListField, fof.DictField):
+            if "subfield" in efield and efield["subfield"] != subfield:
+                efield["subfield"] = None
+            elif subfield is not None:
+                efield["subfield"] = subfield
+        elif is_embedded_doc and etype_is_embedded:
+            # Recursively merge nested embedded document schemas, ensuring the
+            # existing schema is normalized into mapping form before we write
+            # into it.
+            efield["fields"] = _merge_embedded_doc_fields(
+                _to_field_mapping(efield.get("fields")),
+                field.get("fields", []),
+            )
 
-    field["fields"] = fields_dict
+    return merged
 
 
 def _finalize_embedded_doc_fields(fields_dict):
+    """Converts merged embedded-document field specs into list form.
+
+    The intermediate aggregation step stores nested fields in dictionaries so
+    that repeated traversals can be merged efficiently. This helper normalizes
+    the structure back into the list-of-field-specs format expected by
+    `create_field`, recursively applying the same conversion to any nested
+    embedded-document entries.
+    """
+    if not fields_dict:
+        return []
+
+    if isinstance(fields_dict, list):
+        # Older callers may still provide a list that already filtered out
```

Comment:
nit: we don't need to support backwards-compatibility in these private methods. We can assume they are strictly invoked by `get_implied_field_kwargs()`